keps/sig-network/20180612-ipv4-ipv6-dual-stack.md

```diff
@@ -333,36 +336,169 @@ or later. The dual-stack feature requires no changes to this API.
 The versions of CNI plugin binaries that must be used for proper dual-stack functionality (and IPv6 functionality in general) depend upon the version of the container runtime that is used in the cluster nodes (see [CNI issue #531](https://github.com/containernetworking/cni/issues/531) and [CNI plugins PR #113](https://github.com/containernetworking/plugins/pull/113)):
 
 ### Services
+Services depend on `--services-cluster-ip-range` for `ClusterIP` assignment. A dualstack cluster can have this flag configured as `--services-cluster-ip-rang=<cidr>,<cidr>` or `--services-cluster-ip-rang=<cidr>`. The following rules apply:
+
+1. The maximum allowed number of cidrs in this flag is two. Attempting to use more than two cidrs will generate failures during startup. 
+2. When configured with two cidrs then the values are expected to have different ip families. e.g `--services-cluster-ip-range=<ipv4-cidr>,<ipv6-cidr>` or `--services-cluster-ip-range=<ipv6-cidr>,<ipv4-cidr>`. Attempting to configure a cluster with two similar cidrs will generate failures during startup.
+3. A dualstack cluster can have a single cidr in this flag. That means the cluster will be to egress traffic on both ip families, but will *not* be ingress traffic on both ip families.
+4. When upgrading a cluster to dualstack, the first element of the `--services-cluster-ip-range` flag MUST NOT change IP families.  As an example, an IPv4 cluster can safely be upgraded to use `--services-cluster-ip-range=<ipv4-cidr>,<ipv6-cidr>`, but can not safely be changed to `--services-cluster-ip-range=<ipv6-cidr>,<ipv4-cidr>`.
+5. The first cidr entry of `--services-cluster-ip-range=<first cidr entry>, <secondary cidr entry>` is considered cluster default address family.
+6. Attempting to allocate `ClusterIP` from an ip address family that is not configured on the cluster will result in an error.
+
+> The below discussion assumes that `Endpoint Slice and Endpoint Slice Controller`  will allow dual stack endpoints, the existing endpoint controller will remain single stack. That means enabling dual stack services by having more than one entry in `--services-cluster-ip-range` without enabling `Endpoint Slice` feature gate will result in validation failures during startup.
```

Comment:
What will dual stack support look like for EndpointSlices? Although we initially had a single IP type that would have enabled EndpointSlices to store both v4 and v6 addresses, we moved to IP family specific address types for each slice (IPv4 and IPv6). The more generic IP address type is considered deprecated. It seems like the only way to support Pods with v4 and v6 addresses would be entries in two separate EndpointSlices, one for v4 and one for v6. Is that what we want long term?

Reply:
my understanding of the slices is that a Service already can have multiple EndpointSlices, do you foresee any problem using different slices per family?

Reply:
Not an issue per se, just unfortunate if there are consumers that might care about both v4 and v6 addresses for the same endpoint. With the current approach, a Pod that has v4 and v6 addresses will be represented in different EndpointSlices instead of as a single endpoint with 2 addresses in a single EndpointSlice. I just wanted to make that distinction since this almost made it sound like EndpointSlice resources could be dual-stack and not just per-family. That used to be true, but is not currently true.

Reply:
If you think it would be useful to EndpointSlice consumers to have them all together, you could undeprecate the generic IP address type. AIUI it was deprecated because the design of the dual-stack KEP at that time made it look like having separate IPv4 and IPv6 types would make more sense, but now that part of the dual-stack KEP is being changed...

Reply:
Yep, that's exactly it. I should've just asked that more directly to begin with. So my real question is - should we move back to dual-stack EndpointSlices with a single IP address type? Here are some pros and cons for that:

Pros:
- It allows dual stack pods to be represented with a single endpoint in an EndpointSlice.
- It simplifies any implementations that might care about both IPv4 and IPv6 addresses.

Cons:
- It would result in a confusing API change/history.
- It adds some complexity to kube-proxy implementations.

Are there any additional pros or cons I'm missing? Just want to make sure we're all on the same page on the future of EndpointSlices + dual stack.

Reply:
dual-stack is not a thing, is a transition mechanism, the more that we "glorify" it the more difficult to get rid of it ... well, just check the history, we started migrating to IPv6 with dual stack back in the 2000s 🙃 
My vote is for current implementation of endpointslices, at the end the Service IPv6 will handle the EndpointSlice IPv6 and the Service IPv4 the EndpointSlice IPv4 ... but I have to confess that I'm not aware of other implementations that make use of EndpointSlices 😅

Reply:
It's not just kube-proxy that might care.
Just as apps can directly watch for an Endpoints and connect directly, apps could watch for EndpointSlices and, again, make that connection directly (and those apps need to be prepared for there to be more than one EndpointSlice and more than one address family).

Reply:
> dual-stack is not a thing, is a transition mechanism, the more that we "glorify" it the more difficult to get rid of it ... well, just check the history, we started migrating to IPv6 with dual stack back in the 2000s

"Dual-stack is 20 years old, so our 5-year old project can ignore it because it's just a temporary thing"?

At any rate, in a single-stack IPv6 cluster, the "generic IP type" and "IPv6-specific type" solutions would be identical anyway; the difference between them is only noticeable in dual-stack clusters.

Reply:
> "Dual-stack is 20 years old, so our 5-year old project can ignore it because it's just a temporary thing"?

not tried to mean that :smile: , let me clarify, I think that we are trying to solve every possible corner case and I have the feeling that we are running in circles, all the dual-stack discussions in issues, mails and shared docs has more than what , 1000 comments?, any clear outcome? 

My point about the history is that IPv6 migration is a problem that nobody solved in 20 years, dual-stack can be a good solution, but dual-stack is not a problem is a paradox, there are going to be trade-offs always and is impossible to predict what happens next because the reality is that nobody knows.


> Just as apps can directly watch for an Endpoints and connect directly, apps could watch for EndpointSlices and, again, make that connection directly (and those apps need to be prepared for there to be more than one EndpointSlice and more than one address family).

This is an example of what I tried to mean, why the apps can not watch for 2 different endpoint slices, I don't think that all the dual-stack heavy lifting should go on kubernetes implementing an overcomplicated solution just because external apps don't want to watch two 2 endpointslices.

The accepted solution to fix dual-stack brokenness was Happy Eyeballs https://en.wikipedia.org/wiki/Happy_Eyeballs , that's a solution that has to be implemented on clients.
I lean towards that path, implement the right APIs and mechanisms so the clients can adopt the best solution, instead of doing it at the server side

client-go and the API promotion process in kubernetes is the best example, everybody complain about the deprecations and cycles ... but they keep evolving and consumer keep adapting their code :shrug:

Reply:
I think the question is whether it is valuable or not to be able to unambiguously link 2 IP addresses together as "the same backend" across slices.

It's possible to put different ports of the same pod into different slices, too even within the same family, right?

If we think that this is important, we can solve it in a couple different ways.  I am not sure it is important though - do we have any concrete use-cases?

Reply:
> If we think that this is important, we can solve it in a couple different ways. I am not sure it is important though - do we have any concrete use-cases?

@thockin can you ask CoreDNS folks? 
I think they are an important stakeholder and this will have a big impact, I'd really interested in hearing their opinion

Reply:
Do we envision a future where it's common for all Services to be dual stack and all Pods to have v4 and v6 addresses? If so, moving EndpointSlices back to supporting dual stack addresses within each slice could result in half the number of API resources and maybe more importantly, half the number of updates that would need to be transmitted to kube-proxy or other consumers. In most cases I don't think those differences will be noticeable, but it does end up limiting one of the core goals of EndpointSlices - increased scalability. This on its own is likely not worth an API change, just wanted to add that extra factor to this conversation. @wojtek-t does this seem like something that would have any kind of noticeable affect on any scalability efforts?

Reply:
> If so, moving EndpointSlices back to supporting dual stack addresses within each slice could result in half the number of API resources and maybe more importantly, half the number of updates that would need to be transmitted to kube-proxy or other consumers

that's really a good point, the trade off here is number vs size of objects, right? 
I mean, you are always going to have double of IPs, moving back to support dualstack address means half of of objects but double of size.
I remember you did an exhaustive testing on this, I'm really curious now about this 🤔

Reply:
Although we'll have double the IPs, the API objects will not be double in size. Since `addresses` are plural in EndpointSlices, we'd just add an IP address to the endpoint and not have to repeat all the other information such as target ref, topology, hostname, and conditions.

Reply:
> I think the question is whether it is valuable or not to be able to unambiguously link 2 IP addresses together as "the same backend" across slices.

StatefulSets (for example) often need to differentiate specific members of the statefulset with stable identifiers - which is why we have the special `foo-$ordinal` statefulset headless service thing.
I think it is necessary for a dualstack `foo-$ordinal` DNS entry to resolve to the A + AAAA address for the _same_ replica.  So we need enough correlation at the Service/EndpointSlice/whatever to make that possible.

I think dualstack loadbalancer-controller (or ingress-controller?) will _want_ to know that they can (eg) healthcheck on $ip4 but also mark $ip6 unhealthy - or that sending a query to $ip4 counts against the budget of $ip6 too.  That's a somewhat contrived requirement, however.

Reply:
From scalability/performance POV, having a single EndpointSlice object covering IPs of both families seems better. But as also mentioned above, it's not a conceptual difference - it's a matter of a constant, so I wouldn't put too much value to this argument.
I would say that "if all other arguments put both solutions be equally good, then choose the one with single EndpointSlice, if one of those is visibly better for other reasons, go for it".

Reply:
I'd like to ask a question, taking the kube-proxy as an example, please correct me if my reasoning is wrong.

Kube-proxy is using the meta-proxier to handle dual-stack, driving each IP family independently with a different `Proxier`, it would not be better using an endpointSlice per IP family so each `Proxier` can get its endpoints without having to process the endpoint and parse all the IPs locally? (side question, is it possible to do the endpointSlice IP family filtering in the API?) 

I see as an advantage that each IP family is independent, and meta-proxier (thanks @uablrek ) is a good example of how to do an endpoint consumer dual-stack very easily.

Reply:
The main advantage of splitting to different slices:  Easier consumption.  I don't have to "taste" each address to figure out if it is v4 or v6.  This is significant when you have EVERY consumer of EPSlice having to do this.  E.g. Every kube-proxy has to parse every IP address at least once (and currently twice, if I recall) to decide which proxier should act on it.  This seems like a tax we will impose on every consumer of EPSlice.  When we made this decision, I seem to recall that this was the dominant consideration.  Someone tell me I am forgetting something?

I don't see a significant use-case for linking 2 IPs to the same backend that isn't already practically solved.  DNS still have enough information to assign the same name A vs AAAA.  Healthchecks seem contrived and I would argue really DO want to run on both families.

If we think merging is correct, maybe we want to change `Addresses []string` to `Addresses []struct { IPFamily IPFamily; Address string }`.  That leaves out FQDN types.  Maybe we want something more structiured?

```
AddressType AddressType  // IP or FQDN
Endooints []struct Endpoint{
    Addresses []struct EndpointAddress {
        // One-of
        IP struct IPAddress {
            IPFamily IPFamily
            IP string
        }
        FQDN struct FQDNAddress {
            FQDN string
        }
    }
}
```

it's a much more invasive change, but it gives us one Slice with strong validation (all addresses are IPs, each one declares which family it is), all the endpoint-related metadata shared, and easy consumption.  It also leaves a lot of room for expansion if we ever find an AddressType that needs more data than a single `string`.

For compat we could allow IPv4 and IPv6 types, and just require that all addresses in the slice conform.

On the other hand, it means slices will carry "IPv4" over and over again.  1000 endpoints == 4KB of very redundant data. On the other other hand, 2 slices duplicate all the per-endpoint metadata anyway..  If you assume each Endpoint has 100 bytes of data, 1000 endpoints == 100KB of duplicate data.

We could normalize this even further, but I think we're into diminishing returns.

Rob - I'm going to put this back to you - do you think any of this is worthwhile?